Bind in CellParams for RNN

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34140 [quantization] Use torchbind for Linear and RNN PackedParams
* **#35787 Bind in CellParams for RNN**

Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

Commit:
[WIP] Bind in CellParams for RNN

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "[WIP] Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -1182,20 +1670,37 @@ Tensor rnn_relu_cell(
     Tensor output, hy, cy;
-    lstm_cudnn_stub(_input.device().type(), output, hy, cy, _input, hx, _params, has_biases,
-                    num_layers, dropout_p, train, bidirectional, batch_first);
+    // TODOJAMES
```
seems like this is not done yet?

jamesr66a replied:
I think this whole part should be deleted, afaik CUDNN never supported quantized LSTMs

dzhulgakov replied:
then delete it and add TORCH_INTERNAL_ACCEPT(false) or just kill the entire conditional

dzhulgakov reviewed (commented):
RNNs are magical

Commit:
Update on "Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1386,6 +1386,12 @@ struct getTypePtr_<int64_t> final {
 template <>
+struct getTypePtr_<c10::ScalarType> final {
+  static TypePtr call() {
+    return IntType::get();
```
wait, what is it? isn't scalar like a union of int and float?

jamesr66a replied:
Yeah you're right, I think this should be `ScalarTypeType::get()`

jamesr66a replied:
Looks like I have to fix up other existing erroneous sites also, e.g. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_type_parser.cpp#L39

eellison replied:
You're thinking of `Scalar` which is `NumberType::get()`

jamesr66a replied:
also, this is `ScalarType`, not `Scalar`, lol this is confusing

dzhulgakov commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -1182,20 +1670,37 @@ Tensor rnn_relu_cell(
   auto result_dtype = dtype.has_value() ? dtype.value() : at::kChar;
-  check_device(_input, _params, hx);
+  // TODOJAMES
+  // check_device(_input, _params, hx);
```
even if it's always cpu - either just comment out or delete

dzhulgakov reviewed (approved):
Ok, I didn't read each and every line but overall flow seems reasonable and I'll trust you on BC concerns.

Commit:
Update on "Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]

Commit:
Update on "Bind in CellParams for RNN"


Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

[ghstack-poisoned]
